File: maze.py
# Maze Solver - Daniel Charua & Mauricio Rico - 26/09/18
import numpy as np
import sys
# Expanding python's default recursion limit
sys.setrecursionlimit(999999)

class Maze:
    #constructor with file read from input
    def __init__(self):
        self.path = ""
        tmp = []
        #Reading col & row size
        a, b = input().rstrip().split(' ')
        self.col = int(a)
        self.row =  int(b)
        #Reading Start and End point
        a, b = input().rstrip().split(' ')
        self.start_point = int(a), int(b)
        a, b = input().rstrip().split(' ')
        self.end_point = int(a), int(b)
        #Reading maze
        for i in range(self.row):
            tmp.append(list(map(int, input().rstrip())))
        #passing inverted maze to np arry
        self.maze = np.array(tmp[::-1])
        self.solve(self.start_point, self.path)
        #self.printMaze()

    #Function to print maze for debugging
    def printMaze(self):
        c, r = self.start_point
        self.maze[r][c] = 3
        c, r = self.end_point
        self.maze[r][c] = 4
        print('\nCol: %d, Rows: %d' % (self.col, self.row))
        print('start_point: %s marked with 3' % (self.start_point,))
        print('end_point: %s marked with 4' % (self.end_point,))
        print('Path taken: %s ' % (self.path[::-1]))
        print (self.maze)

    #Recursive funcion to solve maze
    def solve(self, coordinates, path):
        c, r = coordinates
        #Base case, end point found,
        if coordinates == self.end_point:
            return True
        #otherwise if its a wall or visited  return False
        elif self.maze[r][c] == 1:
            return False
        elif self.maze[r][c] == 2:
            return False
        #setting visited
        self.maze[r][c] = 2

        #Exploring first to the right if its possible than down, left and up in that order
        if (c < self.col-1 and self.solve((c+1, r), self.path)):
            self.path +="R"
            return True
        elif (r > 0 and self.solve((c, r-1), self.path)):
            self.path +="D"
            return True
        elif (c > 0 and self.solve((c-1, r), self.path)):
            self.path +="L"
            return True
        elif(r < self.row-1 and self.solve((c, r+1), self.path)):
            self.path +="U"
            return True
        return False

    #function to get path inverted to match actual path taken
    def getPath(self):
        return self.path[::-1]

if __name__ == "__main__":
    maze = Maze()
    print( maze.getPath())

File: C++/maze.cpp
//Maze solver - Daniel Charua A01017419 - Mauricio Rico
#include <iostream>
#include <string>
using namespace std;

class Maze{
public:
  int row;
  int col;
  int *maze;
  int start[2];
  int end[2];
  string path;
  string finalPath;
  //Contructor destructor
  Maze(){ path = ""; finalPath = "";}
  ~Maze(){delete(maze);}

  //function to read file
  void readFile(){
    string  line;
    //Read rows and cols
    cin >> line;
    col = stoi(line);
    cin >> line;
    row = stoi(line);
    maze = (int *) malloc(col * row * sizeof(int));
    //Read start point
    cin >> line;
    start[0] = stoi(line);
    cin >> line;
    start[1] = stoi(line);
    //Read end point
    cin >> line;
    end[0] = stoi(line);
    cin >> line;
    end[1] = stoi(line);
    //Read start maze
    for (int i=0; i < row; i++){
      cin >> line;
      for (int j=0; j< row; j++){
        maze[j + ((row-i-1) * row)] =line[j] -'0';
      }
    }
    solveMaze(start[0], start[1]);
  }

  //function to print maze with path for debugging
  void printMaze(){
    printPath();
    for (int i = 0; i < row; i++){
      for (int j = 0; j <col; j++){
        cout<< maze[j + (i * row)];
      }
      cout<<endl;
    }
  }

  //function used recursivly to solve maze
  bool solveMaze(int c, int r){
    if (c == end[0] && r == end[1]){
        return true;
      }
    else if( maze[c + (r * row)] == 1)
        return false;
    else if(maze[c + (r * row)] == 2)
        return false;
    maze[c + (r * row)] = 2;

    if (c < col-1 && solveMaze(c+1, r)){
        path.append("R");
        return true;
      }
    else if(r > 0 && solveMaze(c, r-1)){
        path.append("D");
        return true;
      }
    else if(c > 0 && solveMaze(c-1, r)){
        path.append("L");
        return true;
      }
    else if(r <row -1  && solveMaze(c, r+1)){
        path.append("U");
        return true;
      }
    return false;
  }

  //function to print path
  void printPath(){
    int size = path.length();
    for (int i = size; i >= 0; i--)
      finalPath+=path[i];
    cout<< finalPath;
  }
};

int main(){
  Maze * m = new Maze();
  m->readFile();
  m->printPath();
  delete m;
  return 0;
}
